### ISSUE TITLE
Facets search by second level subcategory does not work

### ISSUE THREAD

**Issue by FranRomero** [PrestaShop/PrestaShop#29727]
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add screenshots

Hi!
I am trying to make a search by second level subcategory but does not seem to work. What I mean for second level subcategory is the following:

Clothes
-> Men (First level subcategory, pending from Clothes)
----> Child (Second level subcategory, pending from Men)

### Expected behavior

That when I filter on facets by second level subcategory the filter would be added and work properly. Now it seems that it is ignoring second level subcategory filters because they are not being applied right now.

### Steps to reproduce

- Installing a cleaning Prestashop v1.7.8.7 with demo data
- By default, it gives to you only 1 level of subcategory for example Clothes > Men so I have created a new pending category for Men called "Child" and I have added some products to it, similar to the previous schema.
- I have configured "Category filter depth" to 0 on "Faceted Search" module to do not have limit on category filter and added the new category on the template.
- Also indexed all mixing prices, rebuild entire price index, build attributes and features indexes and clear caché from "Faceted Search" module.

After made the configuration I go to the front:
If I go to Clothes section, it let me filter by "Men" but not by "Child". Only lets me filter by first level subcategory, like "Men" or "Women", but if I go to "Men" section, it lets me filter by "Child", because right now "Child" its first level subcategory from "Men", so the problem is only when I try to filter by second level subcategory. For example: when I am on "Clothes" and I try to filter by "Child".

### PrestaShop version(s) where the bug happened

1.7.8.7

### PHP version(s) where the bug happened

7.4

### If your bug is related to a module, specify its name and its version

ps_facetedsearch v3.8.0

**Comment by MhiriFaten** [comment]
Hello @FranRomero ,

I can't seem to understand your issue, see the attached screen shots bellow :

### BO 
![image](https://user-images.githubusercontent.com/111756615/191997538-b48e843b-d732-4580-9d01-a57eb21568e8.png)

### FO
![image](https://user-images.githubusercontent.com/111756615/191997490-75dabffd-93b5-4b16-9259-435864d7af7e.png)

![image](https://user-images.githubusercontent.com/111756615/191997599-641131b1-dd48-4631-ab4c-b79335d65770.png)

Could you please provide screen record :pray: 
Waiting for your feedback.

Thanks !

**Comment by FranRomero** [comment]
Hi @MhiriFaten,

Thanks for the fast response. 

I add a video where you can see that I cannot filter by second level subcategory "Child". This does not happen with Men, or Women that are first level subcategories of "Clothes", though. It also does not happen when I enter on "Men" category and try to filter by "Child", because in this case "Child" is first level subcategory of "Men".

I have recorded all on the video, I expect it is good explained enough. If you need some extra info, let me know.

Video link: https://we.tl/t-zuwZa791F8 (I have uploaded it to WeTransfer because it doesnt let me upload here)

Thank you!

**Comment by FranRomero** [comment]
Hi @MhiriFaten,

Have you been able to reproduce the error? Is there any new about it?

Thank you!

**Comment by HanaRebaiQA** [comment]
Hello @FranRomero 

Thanks for your feedback.
I was able to reproduce your issue using the same steps mentioned in your video  with **PS1778** and **PS1787** when putting `"0" `in the `Category filter depth (0 for no limits, 1 by default) `parameter.

The mentioned video : 

https://user-images.githubusercontent.com/92912932/192755273-28e4eeb0-8d66-42cb-9472-d9243978d66d.mov


Here is my different screen record :

**PS1778 : NOK** :x:

https://user-images.githubusercontent.com/92912932/192755316-707d4104-ae14-4b74-94b9-b10ae58816fd.mp4



**PS1787 : NOK** :x: 

https://user-images.githubusercontent.com/92912932/192755673-9625a4dc-eab4-4338-8053-109ad1f2ce8a.mp4




I'll add this to the debug backlog so that it's fixed.

Please be aware that due to the high number of reported issues, some bugs might take a very long time to be resolved. If this issue is important to you and you cannot wait for it to be fixed on the project's own time, we strongly suggest you consider [hiring a specialist](https://www.prestashop.com/en/experts) to help you. 

If you manage to get the problem fixed, then please contribute the fix back to the project as a [Pull Request](https://devdocs.prestashop.com/1.7/contribute/contribute-pull-requests/). Remember that the more people contribute, the better PrestaShop becomes for everyone.

Thank You!

**Comment by HanaRebaiQA** [comment]
It's related to this issue [#16067 ](https://github.com/PrestaShop/PrestaShop/issues/16067)

**Comment by Hlavtox** [comment]
### Investigation of the issue
- `src/Filters/Converter` 
  - Category is searched in `Category::searchByNameAndParentCategoryId($idLang, $queryFilter, (int) $idCategory)`
  - `$queryFilter` contains the name of the category that it got from the encoded facets, `$idCategory` is the ID of the current category OR the home category.
  - So, if the category is deeper in the tree, it will never be found.

### Solution that pops to mind
- Select the category by name without checking the parent. But...
- That brings us to another issue that is connected with it, there can easily be two categories with the same name.

### Only proper solution
- Use IDs for things in the URL, it will fix:
  - This
  - https://github.com/PrestaShop/PrestaShop/issues/23726
  - https://github.com/PrestaShop/PrestaShop/issues/15635

**Comment by uruk-hai-21** [comment]
In version 3.12.1 I found this comment (modules/ps_facetedsearch/src/Filters/Converter.php line 416):
/*
* This works only for categories that are child of the category we are browsing (or home category).
* Categories deeper in the tree will never be found. This could be fixed by providing a unique ID
* to the URL.
*/

What's mean "This could be fixed by providing a unique ID to the URL"?!

**Comment by Pianczetti** [comment]
I made solution like this, somebody should test it
presta 1.7.8.10
facetedsearch 3.13.2

In classes/Category.php added function

    public static function searchCategoryByNameRecursive($idLang, $categoryName, $idParentCategory){
    // Initially, we only search for children of categories with the idParentCategory recognition
    $categories = Category::searchByNameAndParentCategoryId($idLang, $categoryName, (int) $idParentCategory);

    // We recursively search the children of the categories
    if (!$categories) {
        $childCategories = Category::getChildren((int) $idParentCategory, $idLang);
        foreach ($childCategories as $childCategory) {
            $foundCategory = Category::searchCategoryByNameRecursive($idLang, $categoryName, $childCategory['id_category']);
            if ($foundCategory) {
                return $foundCategory;
            }
        }
    }
    return $categories;
    }

Then in ps_facetedsearch/src/Filters/Converter.php line 424 
change
$categories = Category::searchByNameAndParentCategoryId($idLang, $queryFilter, (int) $idCategory);
on
$categories = Category::searchCategoryByNameRecursive($idLang, $queryFilter, (int) $idCategory);

check it and lmk if it's proper solution

**Comment by fregeac** [comment]
J'ai testé et cela fonctionne correctement.
Merci

**Comment by manon-lecareux** [comment]
Hello, has it been merged ? Cause, I still have the issue. Thanks for return

**Comment by webstartconsulting** [comment]
> I made solution like this, somebody should test it presta 1.7.8.10 facetedsearch 3.13.2
> 
> In classes/Category.php added function
> 
> ```
> public static function searchCategoryByNameRecursive($idLang, $categoryName, $idParentCategory){
> // Initially, we only search for children of categories with the idParentCategory recognition
> $categories = Category::searchByNameAndParentCategoryId($idLang, $categoryName, (int) $idParentCategory);
> 
> // We recursively search the children of the categories
> if (!$categories) {
>     $childCategories = Category::getChildren((int) $idParentCategory, $idLang);
>     foreach ($childCategories as $childCategory) {
>         $foundCategory = Category::searchCategoryByNameRecursive($idLang, $categoryName, $childCategory['id_category']);
>         if ($foundCategory) {
>             return $foundCategory;
>         }
>     }
> }
> return $categories;
> }
> ```
> 
> Then in ps_facetedsearch/src/Filters/Converter.php line 424 change $categories = Category::searchByNameAndParentCategoryId($idLang, $queryFilter, (int) $idCategory); on $categories = Category::searchCategoryByNameRecursive($idLang, $queryFilter, (int) $idCategory);
> 
> check it and lmk if it's proper solution

**Thats work on Prestashop 8.1.1 And PS_Faceted Search v4.0.0 versions!** Thanks!

**Comment by boo-code** [comment]
Still reproduces on ps_facetedsearch **5.0.0** with PrestaShop 9.2, and @Hlavtox's 2023 diagnosis is exactly right - the module still carries the comment describing its own limitation at `src/Filters/Converter.php:454`.

Measured, with a control so the mechanism is not just inferred. Created `Child` under `Men`, so it is a second-level subcategory of `Clothes`, then ran the lookup the converter performs:

```
created 'Child' id=16, parent=Men(4), level_depth=4

CURRENT   searchByNameAndParentCategoryId(lang, 'Child', idCategory=3)   NOTHING FOUND   <- browsing Clothes
CONTROL   same call with idCategory=4                                    id 16           <- browsing Men
```

The control is the point: the identical call finds the category one level up, which is why the filter works for `Men` and `Women` but is silently dropped for `Child`. Nothing errors - the facet simply resolves to no id and disappears from the query.

There is already an open PR for this: PrestaShop/ps_facetedsearch#1163 by @Pianczetti, which replaces the parent-scoped name lookup with one scoped by the nested set bounds of the browsed category. I ran its approach against the same fixture:

```
PR #1163 subtree lookup rooted at Clothes(3)                             id 16
```

so it does resolve the case in this report. Two things about it that seem right to me and are worth noting for whoever reviews it: it batches all requested names into a single query rather than one per filter value, and it faces the ambiguity @Hlavtox raised head on - names are not unique in a tree, so it documents the tie-break it picks (shallowest, then lowest id) rather than pretending the problem does not exist. That remains a heuristic; only ids in the URL remove it, which is the "only proper solution" from the 2023 comment.

So this issue is not waiting on a diagnosis any more, it is waiting on a review of #1163.
